AxonDeepSeg/ads_utils.py

```diff
@@ -107,23 +107,64 @@ def imread(filename):
                                f"file extensions:  '.png', '.tif', '.tiff', '.jpg' and '.jpeg'.")
 
-    # Load image
-    if str(file_ext) in ['.tif', '.tiff']:
-        raw_img = imageio.v2.imread(filename, format='tiff-pil')
-        if len(raw_img.shape) > 2:
-            raw_img = imageio.v2.imread(filename, format='tiff-pil', mode='L')
-    else:
-        raw_img = imageio.v2.imread(filename)
-        if len(raw_img.shape) > 2:
-            raw_img = imageio.v2.imread(filename, mode='L')
-    img = imageio.core.image_as_uint(raw_img, bitdepth=8)
+    props = imageio.v3.improps(filename) # new in v3 - standardized metadata
+    _img = imageio.v3.imread(filename)
+
+    if _img.dtype==np.float32 or _img.dtype==np.float64:
+        if len(_img.shape) > 2:
+            raise IOError(f"Multichannel 32bit and 64bit float images are not supported. Please convert the image to 8bit or 16bit, or make a single channel image.")
+
+    # brute force fall back to support backward compatibility 
+    if '.tif' in file_ext:
+        img = np.expand_dims(
+            imageio.v3.imread(filename, plugin='TIFF-PIL'),
+            axis=-1).astype(np.uint8)
+                
+        if len(img.shape) > 3:
+            img = np.expand_dims(
+                imageio.v3.imread(filename, plugin='TIFF-PIL', as_gray=True),
+                axis=-1).astype(np.uint8)
+
+    # c.f for more details: https://github.com/ivadomed/ivadomed/pull/1297#discussion_r1267563980 and 
+    # https://github.com/ivadomed/ivadomed/pull/1297#discussion_r1267563980 
+            
+    # TIFF is a "wild" format. A few assumptions greatly simplify the code below:
+    # 1. the image is interleaved/channel-last (not planar)
+    # 2. the colorspace is one of: binary, gray, RGB, RGBA (not aliasing ones like YUV or CMYK)
+    # 3. the image is flat (not a volume or time-series)
+    # Note: All of these are trivially true for JPEG and PNG due to limitations of these formats.
+
+    # make grayscale (treats binary as 1-bit grayscale)
+    colorspace_idx = 2
+    if _img.ndim <= colorspace_idx:  # binary or gray
+        pass  # nothing to do
+    elif _img.shape[colorspace_idx] == 2:  # gray with alpha channel
+        _img = _img[:, :, 0]
+    elif _img.shape[colorspace_idx] == 3:  # RGB
+        _img = np.sum(_img * (.299, .587, .114), axis=-1)
+    else:  # RGBA
+        # discards alpha
+        _img = np.sum(_img * (.299, .587, .114, 0), axis=-1)
+    if len(_img.shape) < 3:
+        _img = np.expand_dims(_img, axis=-1)
+
+    img = imageio.core.image_as_uint(_img, bitdepth=8)
+
+    # Remove singleton dimension
+    print(img.ndim)
+    print(img.shape)
+    if img.ndim == 3 and img.shape[-1] == 1:
+       img = img[:, :, 0]
+
     return img
 
 def imwrite(filename, img, format='png'):
     """ Write image.
     """
     # check datatype:
-    if img.dtype == 'float64':
+    conversion_list = ['float64', 'float32', 'float16', 'uint16']
```

Comment:
hmm just a heads up that we use `ads_utils.imwrite()` to write the instance segmentations in #742. If there are more than 255 axons in the image, I think it will use 16 bits instead of 8 (we need a unique ID for every axon). In this case, this conversion would be incompatible.

Since we considered revisiting #742, there would be 2 options. Remove this `uint16` to `uint8` conversion, or save the instance seg without this imwrite function.

Reply:
Yeah I recalled #742, but didn't want to mix both of these issues up.

When I was going through this issue, I found that several of our images were not the expected bit depth, and annoyingly sometimes an image might get saved as 16bit using Photoshop but it really is just 8bit depending on the file type (eg png, it's possible to get a 16 bit B&W image saved but if converted to RGB, it will always be 8bit).

So I really just wanted to nail this issue solidely first, and then refactor as needed to resolve #747 and add the required tests at that time

Reply:
Also, if you check the history of the file, because also we've gone through so many iterations of imread of forcing it to 8 bit when read to then allowing flexible  bit depths to then rolling it back, etc.

These tests in this issue will kinda set the boundary of what's possible in terms of png or tiff and bit depth (as far as I could get them to work through manny iterations with saving using photoshop and/or gimp and then reading it in python using different tools)

Reply:
Also note that before #742 can be merged, we need to handle #668 and #698 (unless, were these done already inside #742? I'll have to check)

Reply:
#742 solves #668 and #698 is somewhat unrelated and already implemented (the stanford model even outputs 5 classes)

Reply:
> So I really just wanted to nail this issue solidely first, and then refactor as needed to resolve https://github.com/axondeepseg/axondeepseg/pull/742 and add the required tests at that time

makes sense!